Wrappers/Python/cil/processors/PaganinProcessor.py

```diff
@@ -111,32 +97,86 @@ class PaganinProcessor(Processor):
+        - :math:`k_x, k_y = \left ( \frac{2\pi p}{N_xW}, \frac{2\pi q}{N_yW} 
+        \right )` where :math:`p` and :math:`q` are co-ordinates in a Fourier 
+        mesh in the range :math:`-N_x/2` to :math:`N_x/2` for an image with 
+        size :math:`N_x, N_y` and pixel size :math:`W`.
+    
+    A generalised form of the Paganin phase retrieval method can be called 
+    using :code:`filter_type='generalised_paganin_method'`, which uses the 
+    form of the algorithm described in [2]
+    
+    .. math:: T(x,y) = -\frac{1}{\mu}\ln\left (\mathcal{F}^{-1}\left (\frac{
+        \mathcal{F}\left ( M^2I_{norm}(x, y,z = \Delta) \right )}{1 - \frac{2
+        \alpha}{W^2}\left ( \cos(Wk_x) + \cos(Wk_y) -2 \right )}  \right )
+        \right )
+    
+    The phase retrieval is valid under the following assumptions
+
+        - used with paraxial propagation-induced phase contrast images which 
+        can be assumed to be single-material locally
+        - using intensity data which has been flat field corrected
+        - and under the assumption that the Fresnel number 
+        :math:`F_N = W^2/(\lambda\Delta) >> 1`
+    
+    To apply a filter to images using the Paganin method, call 
+    :code:`full_retrieval=False`. In this case the pre-scaling and conversion 
+    to absorption is not applied so the requirement to supply flat field 
+    corrected intensity data is relaxed,
+    
+    .. math:: I_{filt} = \mathcal{F}^{-1}\left (\frac{\mathcal{F}\left ( 
+        I(x, y,z = \Delta) \right )}
+        {1 - \alpha\left ( k_x^2 + k_y^2 \right )}  \right )
+
     References
     ---------
     - [1] https://doi.org/10.1046/j.1365-2818.2002.01010.x 
-    - [2] https://iopscience.iop.org/article/10.1088/2040-8986/abbab9
-    - [3] https://henke.lbl.gov/optical_constants/getdb2.html
-    With thanks to Rajmund Mokso for help with the initial implementation of the phase retrieval algorithm
+    - [2] https://henke.lbl.gov/optical_constants/getdb2.html
+    - [3] https://iopscience.iop.org/article/10.1088/2040-8986/abbab9
+    With thanks to Rajmund Mokso (DTU) for help with the initial implementation 
+    of the phase retrieval algorithm
 
     """
-
-    UNIT_LIST = ['m','cm','mm','um']
-    UNIT_MULTIPLIERS = [1, 1e-2, 1e-3, 1e-6]
    
-    def __init__(self, delta = 1, beta = 1e-2, energy = 40000,  full_retrieval=True, filter_type='paganin_method', pad = 0, return_units='cm'):
-
-        if return_units not in self.UNIT_LIST:
-            raise ValueError("Return units {} not recognised, expected one of {}, please update data.geometry.config.units".format(str(return_units),str(self.UNIT_LIST)))
-        else:
-            return_multiplier = 1/(self.UNIT_MULTIPLIERS[self.UNIT_LIST.index(return_units)])
+    def __init__(self, delta=1, beta=1e-2, energy=40000,
+                 energy_units=EnergyUnits.eV,  full_retrieval=True, 
+                 filter_type='paganin_method', pad=0, 
+                 return_units=DistanceUnits.cm):
```

Comment:
What are the implications of introducing DistanceUnits? Are these units multiplied onto the actual floating point numbers in some new datatype and as such potentially affecting computational efficiency? Or solely held as a separate string-like thing that users manually look at to know the units?

Reply:
Hi Jakob, the units will be multiplied onto the actual floating point number so yes there's an additional computational cost. We need to have a discussion about when (or if) that calculation happens because we don't want it to be happening inside any of our algorithms. I've started another pull request to think about this https://github.com/TomographicImaging/CIL/pull/1822. I think the principle generally would be that when the user inputs a value with units, we convert it into some common working units which are used for the rest of the time. So the conversion only happens at the point we get values from the user. 

Some of this functionality has moved into this pull request because we wanted to make sure the user interface for the Paganin class will work if we start using units like this. The current interface will work if the user specifies distance with the enum or with a string so we're not wedded to any decisions.

Reply:
Hi Jakob, as discussed in our meeting - I've reverted some of these changes so the new functionality in units will be made in #1822 . For this PR I am converting units internally and the user just inputs the units as a string